Add a basic implementation of ReadableStream async iterable return method

youennf opened the pull request:
#### 7761dec3dc4d8d6d9450e92079deeb981bafbfe7
<pre>
Add a basic implementation of ReadableStream async iterable return method
<a href="https://rdar.apple.com/165606483">rdar://165606483</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=303303">https://bugs.webkit.org/show_bug.cgi?id=303303</a>

Reviewed by Chris Dumez.

As per <a href="https://webidl.spec.whatwg.org/#js-asynchronous-iterator-prototype-object">https://webidl.spec.whatwg.org/#js-asynchronous-iterator-prototype-object</a>, we add support for the return method,
which allows to do error handling when iterating.

We add a concept to detect whether a native iterator implements returnSteps. If so, we expose the return method.
The returnSteps method returns a Ref&lt;DOMPromise&gt;.

We update JSDOMAsyncIterator to implement <a href="https://webidl.spec.whatwg.org/#js-asynchronous-iterator-prototype-object.">https://webidl.spec.whatwg.org/#js-asynchronous-iterator-prototype-object.</a>
We reuse the settle method to handle both runNextSteps and runReturnSteps.
To disambiguate between the two, the settle method will be called with a parameter for runReturnSteps and without for runNextSteps.

Covered by rebased tests.

Canonical link: <a href="https://commits.webkit.org/303823@main">https://commits.webkit.org/303823@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3dfe10ecc47569da1aacdd1525da9ef1d6f1d99

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133715 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44895 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141288 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/df04e64d-60e5-4b50-acfc-c61540a15843) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135585 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6747 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6088 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/810d6329-2fee-42f2-87ca-891bc2f3baea) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136662 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/119867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/83082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/41aae729-37ed-4de3-b96f-502f734ac579) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/113788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/37984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143936 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5894 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/38566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/5978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/5035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/110851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/116123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20667 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5947 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/5793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69411 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/6038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5901 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
```diff
@@ -369,14 +398,130 @@ JSC::EncodedJSValue JSC_HOST_CALL_ATTRIBUTES JSDOMAsyncIteratorPrototype<JSWrapp
+
+    auto encodedValue = JSC::JSValue::encode(value);
+    JSC::ArgList boundArgs { &encodedValue, 1 };
+    auto onFulFilled = JSC::JSFunction::create(vm, &globalObject, 0, String(), onPromiseFulFilled, JSC::ImplementationVisibility::Public);
```
The case is weird here: `onFulFilled`, likely should be `onFulfilled`

youennf replied:
Fixed

cdumez commented on Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
```diff
@@ -369,14 +398,130 @@ JSC::EncodedJSValue JSC_HOST_CALL_ATTRIBUTES JSDOMAsyncIteratorPrototype<JSWrapp
+
+    auto encodedValue = JSC::JSValue::encode(value);
+    JSC::ArgList boundArgs { &encodedValue, 1 };
+    auto onFulFilled = JSC::JSFunction::create(vm, &lexicalGlobalObject, 0, String(), onPromiseFulFilled, JSC::ImplementationVisibility::Public);
```
Same case issue.

youennf replied:
Fixed

youennf committed:
Add a basic implementation of ReadableStream async iterable return method
rdar://165606483
https://bugs.webkit.org/show_bug.cgi?id=303303

Reviewed by Chris Dumez.

As per https://webidl.spec.whatwg.org/#js-asynchronous-iterator-prototype-object, we add support for the return method,
which allows to do error handling when iterating.

We add a concept to detect whether a native iterator implements returnSteps. If so, we expose the return method.
The returnSteps method returns a Ref<DOMPromise>.

We update JSDOMAsyncIterator to implement https://webidl.spec.whatwg.org/#js-asynchronous-iterator-prototype-object.
We reuse the settle method to handle both runNextSteps and runReturnSteps.
To disambiguate between the two, the settle method will be called with a parameter for runReturnSteps and without for runNextSteps.

Covered by rebased tests.

Canonical link: https://commits.webkit.org/303823@main
